include/ignition/rendering/base/BaseScene.hh

```diff
@@ -296,6 +296,21 @@ namespace ignition
       public: virtual ThermalCameraPtr CreateThermalCamera(
                   const unsigned int _id, const std::string &_name) override;
 
+      // Documentation inherited.
+      public: virtual SegmentationCameraPtr CreateSegmentationCamera() override;
+
+      // Documentation inherited.
+      public: virtual SegmentationCameraPtr CreateSegmentationCamera(
+        const unsigned int _id) override;
+
+      // Documentation inherited.
+      public: virtual SegmentationCameraPtr CreateSegmentationCamera(
+        const std::string &_name) override;
+
+      // Documentation inherited.
+      public: virtual SegmentationCameraPtr CreateSegmentationCamera(
+        const unsigned int _id, const std::string &_name) override;
```

Comment:
It looks like this breaks ABI: https://build.osrfoundation.org/job/ignition_rendering-abichecker-any_to_any-ubuntu_auto-amd64/598/API_5fABI_20report/compat_report.html#Type_Binary_Problems_High

According to the [ABI guide from KDE](https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B):

> You cannot...
> 
> add new virtual functions to non-leaf classes as this will break subclasses. Note that a class designed to be sub-classed by applications is always a non-leaf class.
>
> Reason: the addition of a new virtual function to a class that is non-leaf (that is, there is at least one class deriving from this class) changes the layout of the virtual table (the virtual table is basically a list of function pointers, pointing to the functions that are active in this class level). To accommodate the new virtual, the compiler must add a new entry to this table, but existing derived classes won't know about it and will not have the entry in their virtual tables.

So, unless there's a way for us to achieve this functionality without adding any virtual methods to `BaseScene`, we may need to re-target this for Fortress :grimacing:

Reply:
> the addition of a new virtual function to a class that is non-leaf (that is, there is at least one class deriving from this class) changes the layout of the virtual table

I actually believe that the ABI breaking code are the pure virtual methods added to the `Scene` class, since the `Scene` class is a non-leaf class (I accidentally commented on the `BaseScene` code, which could be considered a leaf class):

https://github.com/ignitionrobotics/ign-rendering/blob/3b8c51ea827996f56331ef7ff08b45f67ccc78ed/include/ignition/rendering/Scene.hh#L660-L687

Reply:
ok yeah these will break ABI and It would be difficult to work around it. I recommend targeting at `main`

Reply:
I went ahead and updated this PR to target `main` and updated the segmentation example `CMakeLists.txt` to depend on `ign-rendering6` in https://github.com/ignitionrobotics/ign-rendering/pull/329/commits/8813531a2df28210d292445afc4f5081806b3f8d: https://github.com/ignitionrobotics/ign-rendering/blob/8813531a2df28210d292445afc4f5081806b3f8d/examples/segmentation_camera/CMakeLists.txt#L3

I believe everything should be up-to-date for Fortress now, but it looks like DCO is giving a lot of complaints: https://github.com/ignitionrobotics/ign-rendering/pull/329/checks?check_run_id=2765358606

I know we can set DCO to manually pass, but there seem to be a lot more failures in the check I linked above than usual. @iche033, any thoughts?

Reply:
not sure how DCO got into that state but it's complaining about commits outside of this PR. You can probably just set DCO to pass on this one.

Reply:
> not sure how DCO got into that state

It looks like it had to do with the fact that I reset the branch of this PR to `main` and committed this change before updating the PR itself to target main. I pushed another commit to address some warnings, and now, DCO seems fine. So, I think we can mark this as resolved :+1: